db/system_keyspace.cc

```diff
@@ -2683,6 +2689,21 @@ static service::query_state& internal_system_query_state() {
     return qs;
 };
 
+future<mutation> system_keyspace::make_auth_version_mutation(api::timestamp_type ts, std::string_view version) {
```

Comment:
Ah, I see that you are calling this function from a generator where you only have access to the timestamp. Too bad.

I saw this convention used in a couple of places, so maybe we should (later) reconsider changing the generator interface so that it passes are reference to the current guard instead of just the timestamp.

Reply:
>This will tell the caller that the mutation needs to be computed

I'm not sure why callee should indicate that to the caller, that's like inserting higher level knowledge down to lower layers.

This function just produces mutation with given timestamp, there is nothing about it which makes it only valid withing given guard - that depends on usage.

Reply:
> > This will tell the caller that the mutation needs to be computed
> 
> I'm not sure why callee should indicate that to the caller, that's like inserting higher level knowledge down to lower layers.
> 
> This function just produces mutation with given timestamp, there is nothing about it which makes it only valid withing given guard - that depends on usage.

From the purely technical point of view - yes, but the "auth_version" key is supposed to be managed from group0. Now, it's easier to misuse it because it is much easier to create a timestamp out of thin air than a group0_guard.

Reply:
although we also create guards in many places, and it's not difficult to craft one having group0 client and then use different one when committing it to raft

One could even argue that if you pass timestamp you know that you get mutation with given timestamp and if as a caller you're collecting mutations you know that either they have different timestamps or equal (and you know you want equal). When you pass a guard you don't know this without looking at function implementation.

Reply:
> although we also create guards in many places, and it's not difficult to craft one having group0 client and then use different one when committing it to raft

I don't think there will be more than one group0 client on a particular node. Maybe we will have similar clients for other raft groups (when we have them), but then their type will most likely be separate.

> 
> One could even argue that if you pass timestamp you know that you get mutation with given timestamp and if as a caller you're collecting mutations you know that either they have different timestamps or equal (and you know you want equal). When you pass a guard you don't know this without looking at function implementation.

As a caller, I want to get a mutation that I can apply using given group0 guard. The fact that the mutation needs to have some particular timestamp IMO is a lower-level detail.

Reply:
> I don't think there will be more than one group0 client on a particular node. 

I mean even within the same client you can release guard (just when it goes out of scope for instance) and take a new one easily.

> As a caller, I want to get a mutation that I can apply using given group0 guard

then I guess you need stronger checking mechanism which would assure that (e.g. maybe something like mutation collector in my other pr :) as it owns the guard)